File: edukate-backend/src/main/java/io/github/sanyavertolet/edukate/backend/services/UserService.java
package io.github.sanyavertolet.edukate.backend.services;

import io.github.sanyavertolet.edukate.backend.repositories.UserRepository;
import io.github.sanyavertolet.edukate.common.UserStatus;
import io.github.sanyavertolet.edukate.common.entities.User;
import io.github.sanyavertolet.edukate.common.notifications.SimpleNotificationCreateRequest;
import io.github.sanyavertolet.edukate.common.services.Notifier;
import io.github.sanyavertolet.edukate.common.utils.AuthUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Collection;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;
    private final Notifier notifier;

    public Mono<User> findUserByAuthentication(Authentication authentication) {
        return AuthUtils.monoId(authentication).flatMap(this::findUserById);
    }

    public Mono<User> saveUser(User user) {
        return userRepository.save(user);
    }

    public Mono<User> findUserById(String userId) {
        return userRepository.findById(userId);
    }

    public Flux<User> findUsersByIds(Collection<String> userIds) {
        return userRepository.findAllById(userIds);
    }

    public Mono<User> findUserByName(String name) {
        return userRepository.findByName(name);
    }

    public Mono<Void> deleteUserById(String id) {
        return userRepository.deleteById(id);
    }

    public Mono<Boolean> hasUserPermissionToSubmit(User user) {
        return Mono.just(UserStatus.ACTIVE.equals(user.getStatus()));
    }

    public Mono<Long> notifyAllUsersWithStatus(String title, String message, UserStatus status) {
        return userRepository.findAllByStatus(status)
                .map(user -> new SimpleNotificationCreateRequest(
                        UUID.randomUUID().toString(),
                        user.getId(),
                        title != null ? title : "edukate-talks",
                        message,
                        "edukate team"
                ))
                .flatMap(req -> notifier.notify(req)
                        .onErrorResume(ex -> {
                            log.warn("Could not notify user {}", req.getTargetUserId(), ex);
                            return Mono.empty();
                        })
                )
                .count();
    }
}

File: edukate-backend/src/main/kotlin/io/github/sanyavertolet/edukate/backend/services/UserService.kt
package io.github.sanyavertolet.edukate.backend.services

import io.github.sanyavertolet.edukate.backend.entities.User
import io.github.sanyavertolet.edukate.backend.repositories.UserRepository
import io.github.sanyavertolet.edukate.common.notifications.SimpleNotificationCreateRequest
import io.github.sanyavertolet.edukate.common.services.Notifier
import io.github.sanyavertolet.edukate.common.users.UserStatus
import io.github.sanyavertolet.edukate.common.utils.orNotFound
import java.time.Instant
import java.util.UUID
import org.slf4j.LoggerFactory
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.Caching
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono

@Service
class UserService(private val userRepository: UserRepository, private val notifier: Notifier) {
    @Caching(
        evict =
            [
                CacheEvict(cacheNames = ["users-by-id"], key = "#user.id", condition = "#user.id != null"),
                CacheEvict(cacheNames = ["users-by-name"], key = "#user.name"),
            ]
    )
    fun saveUser(user: User): Mono<User> = userRepository.save(user)

    @Cacheable(cacheNames = ["users-by-id"], key = "#userId")
    fun findUserById(userId: Long): Mono<User> = userRepository.findById(userId)

    @Cacheable(cacheNames = ["users-by-name"], key = "#name")
    fun findUserByName(name: String): Mono<User> = userRepository.findByName(name)

    fun findUserByEmail(email: String): Mono<User> = userRepository.findByEmail(email)

    @Caching(
        evict =
            [
                CacheEvict(cacheNames = ["users-by-id"], key = "#id"),
                CacheEvict(cacheNames = ["users-by-name"], allEntries = true),
            ]
    )
    fun deleteUserById(id: Long): Mono<Void> = userRepository.deleteById(id)

    fun getUserNamesByPrefix(prefix: String, limit: Int, excludeIds: Set<Long> = emptySet()): Flux<String> =
        userRepository
            .findByNamePrefix(prefix, limit + excludeIds.size)
            .filter { it.id !in excludeIds }
            .take(limit.toLong())
            .map { it.name }

    fun hasUserPermissionToSubmit(user: User): Mono<Boolean> = (user.status == UserStatus.ACTIVE).toMono()

    @Caching(
        evict =
            [
                CacheEvict(cacheNames = ["users-by-id"], key = "#userId"),
                CacheEvict(cacheNames = ["users-by-name"], allEntries = true),
            ]
    )
    fun updateName(userId: Long, newName: String): Mono<User> =
        userRepository
            .findByName(newName)
            .flatMap<User> {
                Mono.error(ResponseStatusException(HttpStatus.CONFLICT, "Username '$newName' is already taken"))
            }
            .switchIfEmpty(
                userRepository.findById(userId).orNotFound("User not found").flatMap {
                    userRepository.save(it.copy(name = newName))
                }
            )

    @Caching(
        evict =
            [
                CacheEvict(cacheNames = ["users-by-id"], key = "#userId"),
                CacheEvict(cacheNames = ["users-by-name"], allEntries = true),
            ]
    )
    fun updateEncodedPassword(userId: Long, encodedPassword: String): Mono<Void> =
        userRepository
            .findById(userId)
            .orNotFound("User not found")
            .flatMap { userRepository.save(it.copy(token = encodedPassword)) }
            .then()

    @Caching(
        evict =
            [
                CacheEvict(cacheNames = ["users-by-id"], key = "#userId"),
                CacheEvict(cacheNames = ["users-by-name"], allEntries = true),
            ]
    )
    fun setAvatarTimestamp(userId: Long, instant: Instant?): Mono<User> =
        userRepository.findById(userId).orNotFound("User not found").flatMap {
            userRepository.save(it.copy(avatarUpdatedAt = instant))
        }

    fun notifyAllUsersWithStatus(title: String?, message: String, status: UserStatus): Mono<Long> =
        userRepository
            .findAllByStatus(status)
            .map { user ->
                SimpleNotificationCreateRequest(
                    UUID.randomUUID().toString(),
                    requireNotNull(user.id),
                    title ?: "edukate-talks",
                    message,
                    "edukate team",
                )
            }
            .flatMap { req ->
                notifier.notify(req).onErrorResume { ex ->
                    log.warn("Could not notify user {}", req.targetUserId, ex)
                    Mono.empty()
                }
            }
            .count()

    companion object {
        private val log = LoggerFactory.getLogger(UserService::class.java)
    }
}
